libraries/AP_ADSB/AP_ADSB_Sagetech.cpp

```diff
@@ -0,0 +1,539 @@
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_ADSB_Sagetech.h"
+
+#if HAL_ADSB_SAGETECH_ENABLED
+#include <GCS_MAVLink/GCS.h>
+#include <AP_AHRS/AP_AHRS.h>
+#include <AP_RTC/AP_RTC.h>
+#include <AP_HAL/utility/sparse-endian.h>
+#include <stdio.h>
+#include <time.h>
+#include <string.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define SAGETECH_SCALER_LATLNG              (1.0f/2.145767E-5f)     //   180/(2^23)
+#define SAGETECH_SCALER_KNOTS_TO_CMS        ((KNOTS_TO_M_PER_SEC/0.125f) * 100.0f)
+#define SAGETECH_SCALER_ALTITUDE            (1.0f/0.015625f)
+#define SAGETECH_SCALER_HEADING_CM          ((360.0f/256.0f) * 100.0f)
+
+#define SAGETECH_VALIDFLAG_LATLNG           (1U<<0)
+#define SAGETECH_VALIDFLAG_ALTITUDE         (1U<<1)
+#define SAGETECH_VALIDFLAG_VELOCITY         (1U<<2)
+#define SAGETECH_VALIDFLAG_GND_SPEED        (1U<<3)
+#define SAGETECH_VALIDFLAG_HEADING          (1U<<4)
+#define SAGETECH_VALIDFLAG_V_RATE_GEO       (1U<<5)
+#define SAGETECH_VALIDFLAG_V_RATE_BARO      (1U<<6)
+#define SAGETECH_VALIDFLAG_EST_LATLNG       (1U<<7)
+#define SAGETECH_VALIDFLAG_EST_VELOCITY     (1U<<8)
+
+// detect if any port is configured as Sagetech
+bool AP_ADSB_Sagetech::detect()
+{
+    return (AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_Sagetech_ADSB, 0) != nullptr);
+}
+
+bool AP_ADSB_Sagetech::init()
+{
+    if (_port == nullptr) {
+        _port = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_Sagetech_ADSB, 0);
+    }
+
+    return (_port != nullptr);
+}
+
+void AP_ADSB_Sagetech::update()
+{
+    if (_port == nullptr) {
+        return;
+    }
+
+    const uint32_t now_ms = AP_HAL::millis();
+
+    // -----------------------------
+    // read any available data on serial port
+    // -----------------------------
+    uint32_t nbytes = MIN(_port->available(), 10 * PAYLOAD_XP_MAX_SIZE);
+    while (nbytes-- > 0) {
+        const uint8_t data = (uint8_t)_port->read();
+        if (parse_byte_XP(data)) {
+            handle_packet_XP(message_in_xp.packet);
+        }
+    } // while nbytes
+
+
+    // -----------------------------
+    // handle timers for generating data
+    // -----------------------------
+    if (!last_packet_initialize_ms || (now_ms - last_packet_initialize_ms >= 5000)) {
+        last_packet_initialize_ms = now_ms;
+        send_packet(MsgType_XP::Installation_Set);
+
+
+    } else if (!last_packet_PreFlight_ms || (now_ms - last_packet_PreFlight_ms >= 8200)) {
+        // send once, for now..
+        last_packet_PreFlight_ms = now_ms;
+        // TODO: allow callsign to not require a reboot
+        send_packet(MsgType_XP::Preflight_Set);
+
+    } else if (now_ms - last_packet_Operating_ms >= 1000 && (
+            last_packet_Operating_ms == 0 || // send once at boot

```

Comment:
This doesn't look right to me; should still be a periodic update in case the device powers up late.